src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.Tasks/AzureDevOpsClient.cs

```diff
@@ -0,0 +1,193 @@
+            List<string> sourcePackagePaths = Directory.GetFiles(extractedPackagesPath, "*.nupkg", SearchOption.AllDirectories).ToList();
+        
+            string packageOutputPath = Path.Combine(outputFolder, _packagesFolderName);
+            string symbolOutputPath = Path.Combine(outputFolder, _assetsFolderName);
+
+            if (!Directory.Exists(packageOutputPath))
+            {
+                Directory.CreateDirectory(packageOutputPath);
+            }
+            if (!Directory.Exists(symbolOutputPath))
+            {
+                Directory.CreateDirectory(symbolOutputPath);
+            }
+
+            CopyFiles(packageNames, sourcePackagePaths, extractedPackagesPath, packageOutputPath, flatCopy);
+            CopyFiles(symbolNames, sourceSymbolPaths, extractedAssetsPath, symbolOutputPath, flatCopy);
+        }
+        finally
+        {
+            if (Directory.Exists(extractPath))
+            {
+                Directory.Delete(extractPath, true);
+            }
+            if (File.Exists(downloadPath))
+            {
+                File.Delete(downloadPath);
+            }
+        }
+    }
+
+    private async Task DownloadArtifactZip(string buildId, string artifactName, string downloadPath) 
+    {
+        string relativeUrl = $"build/builds/{buildId}/artifacts?artifactName={artifactName}&api-version={_azureDevOpsApiVersion}";
+        _logger.LogMessage(MessageImportance.High, $"Downloading artifact information from {relativeUrl}");
+        HttpResponseMessage response = await _httpClient.GetAsync(relativeUrl);
+
+        if (!response.IsSuccessStatusCode)
+        {
+            throw new HttpRequestException($"Failed to download artifact information. Status code: {response.StatusCode} Reason: {response.ReasonPhrase}");
+        }
+
+        AzureDevOpsArtifactInformation azdoArtifactInformation = await response.Content.ReadFromJsonAsync<AzureDevOpsArtifactInformation>()
+            ?? throw new ArgumentException($"Couldn't parse AzDo response {response.Content} to {nameof(AzureDevOpsArtifactInformation)}");
+
+        _logger.LogMessage(MessageImportance.High, $"Downloading artifact zip from {azdoArtifactInformation.Resource.DownloadUrl}");
+        response = await _httpClient.GetAsync(azdoArtifactInformation.Resource.DownloadUrl, HttpCompletionOption.ResponseHeadersRead);
+
+        if (!response.IsSuccessStatusCode)
+        {
+            throw new HttpRequestException($"Failed to download artifact zip. Status code: {response.StatusCode} Reason: {response.ReasonPhrase}");
+        }
+
+        using Stream readStream = await response.Content.ReadAsStreamAsync();
+        using FileStream writeStream = File.Create(downloadPath);
+
+        await readStream.CopyToAsync(writeStream, _downloadBufferSize);
+    }
+
+    private void CopyFiles(List<string> fileNamesToCopy, List<string> sourceFiles, string sourceDirectory, string destinationFolder, bool flatCopy)
```

Comment:
The need/purpose of this method isn't super obvious. Can you add a comment as to why this is written this way?